File: projects/payme/java/domain/repository/UserRepository.java
package jk.uz.pdp.domain.repository;

import jk.uz.pdp.domain.exception.DataNotFoundException;
import jk.uz.pdp.domain.exception.EPAMPaymentServerException;
import jk.uz.pdp.domain.model.User;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public abstract class UserRepository implements BaseCrudRepository<User, UUID>{
    private final List<User> userList = new ArrayList<>();

    @Override
    public void create(User user) {
        if(isExist(user.getPhoneNumber())) {
            throw new EPAMPaymentServerException("phone number already registered");
        }
        user.setId(UUID.randomUUID());
        userList.add(user);
    }

    @Override
    public User getById(UUID id) {
        for (User user: userList) {
            if(user.getId().equals(id)) {
                return user;
            }
        }
        throw new DataNotFoundException("User with id: " + id + " not found");
    }

    @Override
    public List<User> getList() {
        return userList;
    }

    @Override
    public User update(User update) {
        int i = 0;
        for (User user: userList) {
            if (user.getId().equals(update.getId())) {
                userList.set(i, update);
                return update;
            }
            i++;
        }
        throw new DataNotFoundException("User with id: " + update.getId() + " not found");
    }

    @Override
    public void delete(UUID id) {
        userList.removeIf(user -> user.getId().equals(id));
        throw new DataNotFoundException("User with id: " + id + " not found");
    }

    public User getUserByPhoneNumber(String phoneNumber) {
        for(User user:userList) {
            if(user.getPhoneNumber().equals(phoneNumber)) {
                return user;
            }
        }
        throw new DataNotFoundException("User with phone number: " + phoneNumber + " not found");
    }

    private boolean isExist(String phoneNumber) {
        for (User user: userList) {
            if(user.getPhoneNumber().equals(phoneNumber)) {
                return true;
            }
        }
        return false;
    }
}

File: projects/payme/typescript/src/repository/user-repository.ts
import { User } from "../entities/user";

type TChecker = (user: User, idx: number) => boolean;

export class UserRepository {
	private list: User[] = [];
	private id = 0;

	create(...users: User[]): void {
		for (let user of users) {
			if (this.isExist(user.phoneNumber)) {
				throw new Error(`User(${user.phoneNumber}) already exists`);
			}

			user.setId(++this.id);
			this.list.push(user);
		}
	}

	getById(userID: number): User {
		for (let user of this.list) {
			if (user.getId() === userID) return user;
		}

		throw new Error(`User(${userID}) not found`);
	}

	getList() {
		return this.list;
	}

	update(update: User) {
		const idx = this.findIndex((user) => user.getId() === update.getId());
		if (idx === undefined) throw new Error("User not found");
		this.list[idx] = update;
	}

	findIndex(fun: TChecker): number | undefined {
		for (let idx = 0; idx < this.list.length; idx++) {
			if (fun(this.list[idx], idx)) {
				return idx;
			}
		}
	}

	delete(id: number) {
		const idx = this.findIndex((user, idx) => user.getId() === id);
		if (idx === undefined) throw new Error("User not found");
		this.list.splice(idx, 1);
	}

	getUserByPhoneNumber(phoneNumber: string): User {
		for (let user of this.list) {
			if (user.phoneNumber === phoneNumber) {
				return user;
			}
		}

		throw new Error(`user(${phoneNumber}) not found`);
	}

	isExist(phoneNumber: string): boolean {
		for (let user of this.list) {
			if (user.phoneNumber === phoneNumber) return true;
		}

		return false;
	}
}
